src/coreComponents/physicsSolvers/wavePropagation/ElasticWaveEquationSEMKernel.hpp

```diff
@@ -0,0 +1,717 @@
+   * @param faceNormal array containing the normal of all faces
+   * @param faceCenter array containing the center of all faces
+   * @param elemsToFaces map to get the global faces from element index and local face index
+   * @param coords coordinate of the point
+   * @return true if coords is inside the element
+   */
+
+  GEOSX_HOST_DEVICE
+  static bool
+  locateSourceElement( real64 const numFacesPerElem,
+                       real64 const (&elemCenter)[3],
+                       arrayView2d< real64 const > const faceNormal,
+                       arrayView2d< real64 const > const faceCenter,
+                       arraySlice1d< localIndex const > const elemsToFaces,
+                       real64 const (&coords)[3] )
+  {
+    //Loop over the element faces
+    real64 tmpVector[3]{};
+    for( localIndex kfe = 0; kfe < numFacesPerElem; ++kfe )
+    {
+
+      localIndex const iface = elemsToFaces[kfe];
+      real64 faceCenterOnFace[3] = {faceCenter[iface][0],
+                                    faceCenter[iface][1],
+                                    faceCenter[iface][2]};
+      real64 faceNormalOnFace[3] = {faceNormal[iface][0],
+                                    faceNormal[iface][1],
+                                    faceNormal[iface][2]};
+
+      //Test to make sure if the normal is outwardly directed
+      LvArray::tensorOps::copy< 3 >( tmpVector, faceCenterOnFace );
+      LvArray::tensorOps::subtract< 3 >( tmpVector, elemCenter );
+      if( LvArray::tensorOps::AiBi< 3 >( tmpVector, faceNormalOnFace ) < 0.0 )
+      {
+        LvArray::tensorOps::scale< 3 >( faceNormalOnFace, -1 );
+      }
+
+      // compute the vector face center to query point
+      LvArray::tensorOps::subtract< 3 >( faceCenterOnFace, coords );
+      localIndex const s = computationalGeometry::sign( LvArray::tensorOps::AiBi< 3 >( faceNormalOnFace, faceCenterOnFace ));
+
+      // all dot products should be non-negative (we enforce outward normals)
+      if( s < 0 )
+      {
+        return false;
+      }
+
+    }
+    return true;
+  }
+
+  /**
+   * @brief Convert a mesh element point coordinate into a coordinate on the reference element
+   * @tparam FE_TYPE finite element type
+   * @param[in] coords coordinate of the point
+   * @param[in] elemsToNodes map to obtaint global nodes from element index
+   * @param[in] X array of mesh nodes coordinates
+   * @param[out] coordsOnRefElem to contain the coordinate computed in the reference element
+   * @return true if coords is inside the element num index
+   */
+  template< typename FE_TYPE >
+  GEOSX_HOST_DEVICE
+  static void
+  computeCoordinatesOnReferenceElement( real64 const (&coords)[3],
+                                        arraySlice1d< localIndex const, cells::NODE_MAP_USD - 1 > const elemsToNodes,
+                                        arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const X,
+                                        real64 (& coordsOnRefElem)[3] )
```

Comment:
Same here, would it be possible to reuse the function also called `computeCoordinatesOnReferenceElement` in `AcousticWaveEquationSEMKernel.hpp`? If needed, you can create a file called `WaveEquationBaseKernel.hpp` that will contain the functions that are common to the two solvers.

Reply:
I'll check it can be a part of our refactoring to